### PR TITLE
Support callable subcommands

### PR DESCRIPTION
**Pull request by ankostis**
Subcommands might be classes, strings (name of classes) and now Callables with a single argument: the parent Application instance (i.e. to read configs from).

That way it is possible to create different subcommands based on the same Application-class just with different constructor arguments.

I could not fins any test-cases about subcommands, so this PR is not considered finished - submitted for discussion.

### CODE REVIEW

File: traitlets/config/application.py [hunk 1]
```diff
@@ -246,11 +246,13 @@ def _flags_changed(self, change):
             assert isinstance(value[1], six.string_types), "Bad flag: %r:%s" % (key, value)
 
 
-    # subcommands for launching other applications
-    # if this is not empty, this will be a parent Application
-    # this must be a dict of two-tuples,
-    # the first element being the application class/import string
-    # and the second being the help string for the subcommand
+    #: subcommands for launching other applications
```
Comment (minrk) [inline traitlets/config/application.py]: looks like these colons were added by accident

File: traitlets/config/application.py [hunk 2]
```diff
@@ -246,11 +246,13 @@ def _flags_changed(self, change):
...
+    #: if this is not empty, this will be a parent Application
+    #: this must be a dict of two-tuples,
+    #: the first element is the application class, or and import string
+    #: or a callable with a single argument for the parent returnng
```
Comment (minrk) [inline traitlets/config/application.py]: returnng

File: traitlets/config/application.py [hunk 3]
```diff
@@ -464,13 +466,16 @@ def initialize_subcommand(self, subc, argv=None):
         """Initialize a subcommand with argv."""
         subapp,help = self.subcommands.get(subc)
 
-        if isinstance(subapp, six.string_types):
-            subapp = import_item(subapp)
+        if callable(subapp):
+            self.subapp = subapp(self)
```
Comment (minrk) [inline traitlets/config/application.py]: classes are callable, so this check will catch the common case of subapp being an Application and handle it incorrectly.

Comment (ankostis) [inline traitlets/config/application.py]: Indeed!

Would this be ok?

    if callable(subapp) and not isinstance(subapp, Application):
        ...

Comment (minrk) [inline traitlets/config/application.py]: I would think so, yes.

Comment (ankostis) [inline traitlets/config/application.py]: Editing my self, more like `issubclass(subapp, Application)`...and also avoid invoking `ConfigSingleton.instance()` when factory method supplied.

Comment (minrk) [inline traitlets/config/application.py]: Thanks!

**Pull request merged**